pkg/storaged/client.js

```diff
@@ -206,6 +214,165 @@
+        client.uuids_btrfs_default_subvol = { };
+    if (!client.uuids_btrfs_volume)
+        return;
+
+    if (!client.superuser.allowed) {
+        return;
+    }
+
+    const uuids_subvols = { };
+    const uuids_usage = { };
+    const btrfs_mounts = { };
+    const btrfs_default_subvol = { };
+    for (const uuid of Object.keys(client.uuids_btrfs_volume)) {
+        // In multi device setups MountPoints can be on either of the block devices, so try them all.
+        const blocks = client.uuids_btrfs_blocks[uuid];
+        const MountPoints = blocks.map(block => {
+            return client.blocks_fsys[block.path];
+        }).map(block_fsys => block_fsys.MountPoints).reduce((accum, current) => accum.concat(current));
+        const mp = MountPoints[0];
+        if (mp) {
+            let output;
+            const mount_point = utils.decode_filename(mp);
+            try {
+                output = await cockpit.spawn(["btrfs", "subvolume", "list", "-ap", mount_point], { superuser: true, err: "message" });
+            } catch (err) {
+                console.error(`unable to obtain subvolumes for mount point ${mount_point}`, err);
+                return;
+            }
+            const subvols = [{ pathname: "/", id: 5 }];
+            for (const line of output.split("\n")) {
+                const m = line.match(/ID (\d+).*parent (\d+).*path (<FS_TREE>\/)?(.*)/);
+                if (m)
+                    subvols.push({ pathname: m[4], id: Number(m[1]) });
+            }
+            uuids_subvols[uuid] = subvols;
+
+            // HACK: Obtain the default subvolume, required for mounts in which do not specify a subvol and subvolid.
+            // In the future can be obtained via UDisks, it requires the btrfs partition to be mounted somewhere.
+            // https://github.com/storaged-project/udisks/commit/b6966b7076cd837f9d307eef64beedf01bc863ae
+            try {
+                output = await cockpit.spawn(["btrfs", "subvolume", "get-default", mount_point], { superuser: true, err: "message" });
+            } catch (err) {
+                console.error(`unable to obtain default subvolume for mount point ${mount_point}`, err);
+                return;
+            }
+
+            const id_match = output.match(/ID (\d+).*/);
+            if (id_match)
+                btrfs_default_subvol[uuid] = Number(id_match[1]);
+
+            // HACK: UDisks should expose a better btrfs API with btrfs device information
+            // https://github.com/storaged-project/udisks/issues/1232
+            // TODO: optimise into just parsing one `btrfs filesystem show`?
+            const usages = {};
+            let usage_output;
+            try {
+                usage_output = await cockpit.spawn(["btrfs", "filesystem", "show", "--raw", uuid], { superuser: true, err: "message" });
+            } catch (err) {
+                console.error(`btrfs filesystem show ${uuid}`, err);
+                return;
```

Comment:
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19690-20231211-210241-6861d8fd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)